# **Task 1**
Creating the best prediction model to predict “Churn Status in Banking”.

a. Download “Assignment_04_Attachment_01.ipynb”.

b. Upload on your Google Colab.

c. Complete the code, so that you can get 3 different models.

d. Compare those 3 model evaluations, and summarize which model is
the best. e. Use any Model Optimization Method (hyperparameter
tuning, feature selection, etc.) if needed.

# Preparation

Preparation biasa dilakukan untuk mempersiapkan data sebelum masuk dalam tahap pemodelan. <br>
Berikut adalah tahapan yang akan dilalui pada data `SC_HW1_bank_data.csv` sebelum tahap pemodelan :
1. Import Library
2. Input Dataset
3. Preprocessing
4. Train-Test Split

## Import Library

In [ ]:
import pandas as pd
import numpy as np

## Input Dataset

In [ ]:
#Membaca data dan memasukkannya ke dalam bentuk Pandas Dataframe
df = pd.read_csv('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv')

In [ ]:
#Jalankan code untuk mengecek nama kolom yang tersedia
df.columns

Index(['RowNumber', 'CustomerId', 'Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited'],
      dtype='object')

In [ ]:
# Menampilkan data teratas
df.head()

RowNumber  CustomerId Geography  Gender  Age  Tenure    Balance  \
0          1    15634602    France  Female   42       2       0.00   
1          2    15647311     Spain  Female   41       1   83807.86   
2          3    15619304    France  Female   42       8  159660.80   
3          4    15701354    France  Female   39       1       0.00   
4          5    15737888     Spain  Female   43       2  125510.82   

   NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  Exited  
0              1          1               1        101348.88       1  
1              1          0               1        112542.58       0  
2              3          1               0        113931.57       1  
3              2          0               0         93826.63       0  
4              1          1               1         79084.10       0

In [ ]:
#Hilangkan kolom yang dirasa tidak relevan dengan model (contoh: ID). None dapat diisi dengan nama-nama kolom yang akan digunakan.
#Contoh df = df[['X1','X2', 'Y']].copy()

#START CODE
df = df[['Geography', 'Gender', 'Age', 'Tenure',
       'Balance', 'NumOfProducts', 'HasCrCard', 'IsActiveMember',
       'EstimatedSalary', 'Exited']].copy()
df

Geography  Gender  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0       France  Female   42       2       0.00              1          1   
1        Spain  Female   41       1   83807.86              1          0   
2       France  Female   42       8  159660.80              3          1   
3       France  Female   39       1       0.00              2          0   
4        Spain  Female   43       2  125510.82              1          1   
...        ...     ...  ...     ...        ...            ...        ...   
9995    France    Male   39       5       0.00              2          1   
9996    France    Male   35      10   57369.61              1          1   
9997    France  Female   36       7       0.00              1          0   
9998   Germany    Male   42       3   75075.31              2          1   
9999    France  Female   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  
0                  1        101348.88       1  
1                  1        112542.58       0  
2                  0        113931.57       1  
3                  0         93826.63       0  
4                  1         79084.10       0  
...              ...              ...     ...  
9995               0         96270.64       0  
9996               1        101699.77       0  
9997               1         42085.58       1  
9998               0         92888.52       1  
9999               0         38190.78       0  

[10000 rows x 10 columns]

## Preprocessing

In [ ]:
df = pd.read_csv(('https://raw.githubusercontent.com/Rietaros/kampus_merdeka/main/SC_HW1_bank_data.csv'), dtype={'kolom_kategorikal': 'category'})

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10000 non-null  int64  
 1   CustomerId       10000 non-null  int64  
 2   Geography        10000 non-null  object 
 3   Gender           10000 non-null  object 
 4   Age              10000 non-null  int64  
 5   Tenure           10000 non-null  int64  
 6   Balance          10000 non-null  float64
 7   NumOfProducts    10000 non-null  int64  
 8   HasCrCard        10000 non-null  int64  
 9   IsActiveMember   10000 non-null  int64  
 10  EstimatedSalary  10000 non-null  float64
 11  Exited           10000 non-null  int64  
dtypes: float64(2), int64(8), object(2)
memory usage: 937.6+ KB


In [ ]:
#Lakukan One-Hot Encoder untuk data categorical, dengan fungsi pandas get_dummies

#START CODE
df = pd.get_dummies(df, columns=['Geography', 'Gender'])

In [ ]:
df

RowNumber  CustomerId  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0             1    15634602   42       2       0.00              1          1   
1             2    15647311   41       1   83807.86              1          0   
2             3    15619304   42       8  159660.80              3          1   
3             4    15701354   39       1       0.00              2          0   
4             5    15737888   43       2  125510.82              1          1   
...         ...         ...  ...     ...        ...            ...        ...   
9995       9996    15606229   39       5       0.00              2          1   
9996       9997    15569892   35      10   57369.61              1          1   
9997       9998    15584532   36       7       0.00              1          0   
9998       9999    15682355   42       3   75075.31              2          1   
9999      10000    15628319   28       4  130142.79              1          1   

      IsActiveMember  EstimatedSalary  Exited  Geography_France  \
0                  1        101348.88       1                 1   
1                  1        112542.58       0                 0   
2                  0        113931.57       1                 1   
3                  0         93826.63       0                 1   
4                  1         79084.10       0                 0   
...              ...              ...     ...               ...   
9995               0         96270.64       0                 1   
9996               1        101699.77       0                 1   
9997               1         42085.58       1                 1   
9998               0         92888.52       1                 0   
9999               0         38190.78       0                 1   

      Geography_Germany  Geography_Spain  Gender_Female  Gender_Male  
0                     0                0              1            0  
1                     0                1              1            0  
2                     0                0              1            0  
3                     0                0              1            0  
4                     0                1              1            0  
...                 ...              ...            ...          ...  
9995                  0                0              0            1  
9996                  0                0              0            1  
9997                  0                0              1            0  
9998                  1                0              0            1  
9999                  0                0              1            0  

[10000 rows x 15 columns]

In [ ]:
#Pisahkan mana X (feature) dengan Y,
#Y adalah kolom "Exited"

#START CODE
X = df.drop(columns=['Exited'])  # Take all columns except the 'Exited' column as features
y = df['Exited']  # Take the 'Exited' column as the target

In [ ]:
X.head()

RowNumber  CustomerId  Age  Tenure    Balance  NumOfProducts  HasCrCard  \
0          1    15634602   42       2       0.00              1          1   
1          2    15647311   41       1   83807.86              1          0   
2          3    15619304   42       8  159660.80              3          1   
3          4    15701354   39       1       0.00              2          0   
4          5    15737888   43       2  125510.82              1          1   

   IsActiveMember  EstimatedSalary  Geography_France  Geography_Germany  \
0               1        101348.88                 1                  0   
1               1        112542.58                 0                  0   
2               0        113931.57                 1                  0   
3               0         93826.63                 1                  0   
4               1         79084.10                 0                  0   

   Geography_Spain  Gender_Female  Gender_Male  
0                0              1            0  
1                1              1            0  
2                0              1            0  
3                0              1            0  
4                1              1            0

In [ ]:
y.head()

0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [ ]:
#Lakukan Scaler dan/atau Noermalisasi jika diperlukan
from sklearn.preprocessing import MinMaxScaler

#START CODE
scaler = MinMaxScaler()
X_transform = scaler.fit_transform(X)

In [ ]:
#Ini digunakan jika dilakukan scaler/Normalisas. Jika tidak, code ini bisa dilewat dan diganti dengan code yang ada di dalam komen
X_transform = pd.DataFrame(X_transform, columns = X.columns)
#X_transform = X.copy()

## Train-Test Split

In [ ]:
#Split menjadi train dan test dengan test_size 25%
#Tidak perlu mengubah code ini

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Modeling

# **Model1**

### Soal :
Jelaskan secara Singkat Model pertama yang digunakan!

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
model1 = LogisticRegression()
params = None
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model1,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

## **Jawaban**

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV

# Define the model
model1 = LogisticRegression()

# Define the hyperparameter grid to search
params = {"tol": [0.1, 0.01, 0.001], 'C':[0.5, 1.0, 1.5, 2.0]}

# Create a GridSearchCV object
grid = GridSearchCV(
    estimator=model1,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,  # Number of CPU cores to use
    cv=10  # 10-fold cross-validation
)

# Fit the grid search to your data
grid.fit(X_train, y_train)

# Get the best hyperparameters
best_params = grid.best_params_
best_params

{'C': 1.5, 'tol': 0.01}

In [ ]:
#lakukan evaluasi dengan beberapa maetric di bawah ini
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.82      0.97      0.89      1983
           1       0.64      0.21      0.31       517

    accuracy                           0.81      2500
   macro avg       0.73      0.59      0.60      2500
weighted avg       0.79      0.81      0.77      2500


[[1921   62]
 [ 409  108]]

0.8116


## **Kesimpulan Model Logistic Regression**

Model Logistic Regression adalah model statistik yang digunakan untuk masalah klasifikasi. Cara kerja model Logistic Regression adalah dengan menggunakan fungsi logistik (sigmoid) untuk mengubah hasil dari kombinasi linear fitur-fitur menjadi nilai antara 0 dan 1, yang dapat diinterpretasikan sebagai probabilitas. Jika probabilitas lebih besar dari 0.5, model akan mengklasifikasikan instance ke dalam kelas 1; jika probabilitas lebih kecil atau sama dengan 0.5, instance akan diklasifikasikan ke dalam kelas 0. Hasil dari evaluasi kinerja model Logistic Regression ditampilkan dalam metrik seperti precision, recall, F1-score, dan akurasi.

1. Precision:

> - Precision adalah rasio positif benar (true positive) dibagi dengan total prediksi positif (true positive + false positive).
> - Untuk kelas 0 (label 0), precision adalah 0.82, yang berarti dari semua prediksi yang diklasifikasikan sebagai kelas 0, 82% benar-benar adalah kelas 0.
> - Untuk kelas 1 (label 1), precision adalah 0.64, yang berarti dari semua prediksi yang diklasifikasikan sebagai kelas 1, 64% benar-benar adalah kelas 1.

2. Recall (Sensitivity):

> - Recall adalah rasio positif benar (true positive) dibagi dengan total jumlah data aktual yang sebenarnya adalah kelas positif (true positive + false negative).
> - Untuk kelas 0 (label 0), recall adalah 0.97, yang berarti bahwa dari semua data yang sebenarnya adalah kelas 0, model berhasil mengidentifikasi 97% dari mereka.
> - Untuk kelas 1 (label 1), recall adalah 0.21, yang berarti bahwa dari semua data yang sebenarnya adalah kelas 1, model hanya berhasil mengidentifikasi 21% dari mereka.

3. F1-Score:

> - F1-score adalah rata-rata harmonik dari precision dan recall. Ini memberikan nilai tunggal yang mencakup baik precision maupun recall.
> - F1-score untuk kelas 0 adalah 0.89, dan untuk kelas 1 adalah 0.31.

4. Support:

> Support adalah jumlah data aktual yang termasuk dalam setiap kelas.

5.Accuracy:

> - Accuracy adalah rasio prediksi yang benar (true predictions) dibagi dengan total jumlah prediksi.
> - Accuracy adalah 0.81, yang berarti model Anda memiliki tingkat keakuratan sekitar 81% pada dataset ini.

6. Confusion Matrix:

> - Confusion matrix menunjukkan perbandingan antara prediksi model dengan kelas sebenarnya.
> - Diagonal utama dari matriks ini adalah true positive (1921) dan true negative (108), yang berarti model benar-benar memprediksi dengan benar kelas 0 sebanyak 1921 kali dan kelas 1 sebanyak 108 kali.
> - Diagonal yang tidak utama adalah false positive (62) dan false negative (409), yang berarti model salah memprediksi kelas 0 sebagai kelas 1 sebanyak 62 kali dan kelas 1 sebagai kelas 0 sebanyak 409 kali.

Kesimpulannya, model ini memiliki kinerja yang baik dalam mengklasifikasikan kelas 0 tetapi memiliki kesulitan dalam mengklasifikasikan kelas 1. Diperlukan upaya lebih lanjut, seperti penyesuaian hyperparameter atau penggunaan metode lain, untuk meningkatkan kinerja model terutama pada kelas 1 jika itu adalah tujuan utama dalam konteks aplikasi yang dihadapi.

# **Model2**
### Soal :
Jelaskan secara Singkat Model ke-2 yang digunakan!

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
model2 = None
params = None
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model2,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

## **Jawaban**

In [ ]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV

# Define the model
model2 = RandomForestClassifier()

# Define the hyperparameter grid to search
params = {
    "n_estimators": [50, 100, 200],
    "max_depth": [None, 10, 20, 30],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4],
    "max_features": ["auto", "sqrt", "log2"],
}

# Create a GridSearchCV object
grid = GridSearchCV(
    estimator=model2,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,  # Number of CPU cores to use
    cv=10  # 10-fold cross-validation
)

# Fit the grid search to your data
grid.fit(X_train, y_train)

# Get the best hyperparameters
best_params = grid.best_params_
best_params

{'max_depth': 20,
 'max_features': 'sqrt',
 'min_samples_leaf': 2,
 'min_samples_split': 5,
 'n_estimators': 200}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.97      0.92      1983
           1       0.79      0.45      0.57       517

    accuracy                           0.86      2500
   macro avg       0.83      0.71      0.75      2500
weighted avg       0.85      0.86      0.85      2500


[[1920   63]
 [ 283  234]]

0.8616


## **Kesimpulan Model Random Forest**

Model Random Forest didasarkan pada konsep Decision Tree, tetapi bekerja dengan menggabungkan banyak pohon keputusan (Decision Trees) menjadi satu model yang lebih kuat. Keuntungan utama dari Random Forest adalah kemampuannya untuk mengatasi masalah overfitting yang sering terjadi pada pohon keputusan tunggal. Selain itu, Random Forest biasanya memberikan hasil yang baik dalam berbagai jenis data dan cocok untuk digunakan sebagai model referensi dalam banyak tugas klasifikasi dan regresi.

1. Precision:

> - Precision adalah ukuran sejauh mana model memprediksi kelas positif dengan benar.
> - Untuk kelas 0 (label 0), precision adalah 0.87, yang berarti bahwa dari semua prediksi yang diklasifikasikan sebagai kelas 0, sekitar 87% benar-benar adalah kelas 0.
> - Untuk kelas 1 (label 1), precision adalah 0.79, yang berarti bahwa dari semua prediksi yang diklasifikasikan sebagai kelas 1, sekitar 79% benar-benar adalah kelas 1.

2. Recall (Sensitivity):

> - Recall adalah ukuran sejauh mana model berhasil mengidentifikasi semua kasus yang sebenarnya positif.
> - Untuk kelas 0 (label 0), recall adalah 0.97, yang mengindikasikan bahwa model sangat baik dalam mengidentifikasi data yang sebenarnya adalah kelas 0.
> - Untuk kelas 1 (label 1), recall adalah 0.45, yang menunjukkan bahwa model memiliki kesulitan dalam mengidentifikasi data yang sebenarnya adalah kelas 1.

3. F1-Score:

> - F1-score adalah rata-rata harmonik dari precision dan recall. Ini memberikan nilai tunggal yang mencakup baik precision maupun recall.
> - F1-score untuk kelas 0 adalah 0.92, dan untuk kelas 1 adalah 0.57.

4. Support:

> Support adalah jumlah data aktual yang termasuk dalam setiap kelas. Ada 1983 data dengan label 0 dan 517 data dengan label 1.

5. Accuracy:

> Akurasi model adalah 0.8616, yang berarti model memiliki tingkat keakuratan sekitar 86.16% pada dataset ini.

6. Confusion Matrix:

> - Confusion matrix menunjukkan perbandingan antara prediksi model dengan kelas sebenarnya.
> - Diagonal utama dari matriks ini adalah true positive (1920) dan true negative (234), yang berarti model benar-benar memprediksi dengan benar kelas 0 sebanyak 1920 kali dan kelas 1 sebanyak 234 kali.
> - Diagonal yang tidak utama adalah false positive (63) dan false negative (283), yang mengindikasikan bahwa model melakukan beberapa kesalahan dalam mengklasifikasikan data kelas 1.

Kesimpulannya, model ini memiliki kinerja yang baik dalam mengidentifikasi data kelas 0 dengan tingkat precision dan recall yang tinggi. Namun, model memiliki kesulitan dalam mengidentifikasi data kelas 1, yang tercermin dalam recall yang lebih rendah. Akurasi model secara keseluruhan juga cukup baik.

# **Model3**
### Soal :
Jelaskan secara Singkat Model ke-3 yang digunakan!

In [ ]:
#Pilih salah satu metode Machine Leaarning
#Model Machine Learning dapat dipanggil terlebih dahulu melalui library yang digunakan. Gunakan library scikit learn seperti pada contoh

#Contoh pemanggilan library dan penggunaannya dalam model

#from sklearn.linear_model import LogisticRegression
#model1 = LogisticRegression()
#params = { "tol": [0.1,0.01,0.001], 'C':[0.5,1.0,1.5,2.0]}


#START CODE
model3 = None
params = None
#END CODE

#Lakukan parameter tuning sesuai hyperparameter yang dibutuhkan
from sklearn.model_selection import GridSearchCV
grid = GridSearchCV(
             estimator= model3,
             param_grid= params,
             scoring = 'accuracy',
             n_jobs = 10, # core cpu yang digunakan
             cv = 10 # 3-fold cross validation (artinya kita melakukan iterasi model sebanyak 3 kali)
            )

grid.fit(X_train,y_train)
grid.best_params_

## **Jawaban**

In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import GridSearchCV

# Define the model
model3 = GradientBoostingClassifier()

# Define the hyperparameter grid to search
params = {
    "n_estimators": [50, 100, 200],
    "learning_rate": [0.01, 0.1, 0.2],
    "max_depth": [3, 4, 5],
    "min_samples_split": [2, 5, 10],
    "min_samples_leaf": [1, 2, 4],
}

# Create a GridSearchCV object
grid = GridSearchCV(
    estimator=model3,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,  # Number of CPU cores to use
    cv=10  # 10-fold cross-validation
)

# Fit the grid search to your data
grid.fit(X_train, y_train)

# Get the best hyperparameters
best_params = grid.best_params_
best_params

{'learning_rate': 0.1,
 'max_depth': 5,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'n_estimators': 50}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.88      0.96      0.92      1983
           1       0.77      0.49      0.60       517

    accuracy                           0.86      2500
   macro avg       0.83      0.72      0.76      2500
weighted avg       0.86      0.86      0.85      2500


[[1909   74]
 [ 266  251]]

0.864


## **Kesimpulan Model Gradient Boosting**

Model Gradient Boosting adalah salah satu teknik ensemble learning yang digunakan dalam machine learning untuk tugas klasifikasi dan regresi. Ini adalah model yang kuat yang membangun serangkaian pohon keputusan (biasanya Decision Trees) secara berurutan dengan cara yang cerdas untuk mengoreksi kesalahan prediksi model sebelumnya. Keuntungan utama dari model Gradient Boosting adalah kemampuannya untuk menghasilkan prediksi yang sangat akurat dan kuat, terutama dalam kasus di mana data memiliki kompleksitas yang tinggi. Namun, proses pelatihan bisa menjadi lebih lambat dan membutuhkan penyesuaian parameter yang hati-hati.

1. Precision:

> - Precision mengukur seberapa banyak dari prediksi yang positif yang sebenarnya positif.
> - Untuk kelas 0, precision adalah 0,88, yang berarti sekitar 88% dari semua prediksi yang diklasifikasikan sebagai kelas 0 adalah benar.
> - Untuk kelas 1, precision adalah 0,77, yang mengindikasikan sekitar 77% dari semua prediksi yang diklasifikasikan sebagai kelas 1 adalah benar.

2. Recall (Sensitivity):

> - Recall mengukur seberapa banyak dari instance positif yang sebenarnya positif yang berhasil diprediksi.
> - Untuk kelas 0, recall adalah 0,96, menunjukkan bahwa model sangat baik dalam mengidentifikasi instance kelas 0 yang sebenarnya.
> - Untuk kelas 1, recall adalah 0,49, yang mengindikasikan bahwa model kesulitan dalam mengidentifikasi instance kelas 1 yang sebenarnya.

3. F1-Score:

> - F1-score adalah rata-rata harmonik dari precision dan recall. Ini memberikan nilai tunggal yang menggabungkan baik precision maupun recall.
> - F1-score untuk kelas 0 adalah 0,92, dan untuk kelas 1 adalah 0,60.

4. Support:

> Support mewakili jumlah instance aktual dalam setiap kelas. Ada 1983 instance dengan label 0 dan 517 instance dengan label 1.

5. Akurasi:

> Akurasi model adalah 0,864, yang berarti model secara benar mengklasifikasikan sekitar 86,4% dari total instance.

6. Confusion Matrix:

> - Confusion matrix menunjukkan perbandingan antara prediksi model dengan kelas sebenarnya.
> - Diagonal utama terdiri dari true positive (1909) dan true negative (251), menunjukkan bahwa model dengan benar memprediksi kelas 0 sebanyak 1909 kali dan kelas 1 sebanyak 251 kali.
> - Elemen di luar diagonal adalah false positive (74) dan false negative (266), menunjukkan bahwa model membuat beberapa kesalahan dalam mengklasifikasikan instance kelas 1.

Secara keseluruhan, model ini memiliki kinerja yang baik dalam mengidentifikasi instance kelas 0 dengan precision dan recall yang tinggi. Namun, model kesulitan dalam mengidentifikasi instance kelas 1, seperti yang tercermin dalam recall yang lebih rendah. Akurasi model secara keseluruhan relatif tinggi, tetapi perlu mempertimbangkan trade-off antara precision dan recall tergantung pada tujuan.

# **Model4**
### Soal :
Jelaskan secara Singkat Model ke-4 yang digunakan!

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV

# Define the model
model4 = KNeighborsClassifier()

# Define the hyperparameter grid to search
params = {
    "n_neighbors": [3, 5, 7],
    "weights": ["uniform", "distance"],
    "p": [1, 2]  # 1 for Manhattan distance, 2 for Euclidean distance
}

# Create a GridSearchCV object
grid = GridSearchCV(
    estimator=model4,
    param_grid=params,
    scoring='accuracy',
    n_jobs=10,  # Number of CPU cores to use
    cv=10  # 10-fold cross-validation
)

# Fit the grid search to your data
grid.fit(X_train, y_train)

# Get the best hyperparameters
best_params = grid.best_params_
best_params

{'n_neighbors': 7, 'p': 1, 'weights': 'distance'}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.83      0.96      0.89      1983
           1       0.61      0.26      0.36       517

    accuracy                           0.81      2500
   macro avg       0.72      0.61      0.62      2500
weighted avg       0.79      0.81      0.78      2500


[[1898   85]
 [ 385  132]]

0.812


## **Kesimpulan Model K-Nearest Neighbors**

Model K-Nearest Neighbors (KNN) adalah salah satu model machine learning yang digunakan untuk masalah klasifikasi dan regresi. Ini adalah model yang berbasis instansi (instance-based), yang berarti ia tidak mempelajari fungsi pemetaan dari data pelatihan, tetapi melakukan prediksi berdasarkan kedekatan (proximity) antara instance yang akan diprediksi dengan instance-instance yang ada dalam data pelatihan. Kinerja model KNN dievaluasi dengan metrik-metrik seperti akurasi, precision, recall, dan F1-score.

1. Kelas 0 (label 0):

> - Precision: 0.83, yang menunjukkan bahwa sekitar 83% dari prediksi yang diklasifikasikan sebagai kelas 0 adalah benar-benar kelas 0.
> - Recall: 0.96, yang mengindikasikan bahwa model sangat baik dalam mengidentifikasi data yang sebenarnya adalah kelas 0.
> - F1-Score: 0.89, yang merupakan rata-rata harmonik antara precision dan recall untuk kelas 0.
> - Terdapat 1983 data dengan label 0.

2. Kelas 1 (label 1):

> - Precision: 0.61, yang berarti bahwa sekitar 61% dari prediksi yang diklasifikasikan sebagai kelas 1 adalah benar-benar kelas 1.
> - Recall: 0.26, yang menunjukkan bahwa model memiliki kesulitan dalam mengidentifikasi data yang sebenarnya adalah kelas 1.
> - F1-Score: 0.36, yang merupakan rata-rata harmonik antara precision dan recall untuk kelas 1.
> - Terdapat 517 data dengan label 1.

3. Akurasi (Accuracy):

> Akurasi model adalah 0.812 atau sekitar 81.2%, yang berarti model ini memiliki tingkat keakuratan yang cukup baik secara keseluruhan pada dataset ini.

5. Confusion Matrix:

> - Model memiliki jumlah true positives yang cukup tinggi untuk kelas 0 (1898) dan kelas 1 (132).
> - Terdapat jumlah yang signifikan dari false positives (85) dan false negatives (385), yang mengindikasikan bahwa model ini melakukan beberapa kesalahan dalam mengklasifikasikan data kelas 1.

Kesimpulannya adalah model ini memiliki kinerja yang baik dalam mengidentifikasi data kelas 0, tetapi memiliki kinerja yang kurang baik dalam mengidentifikasi data kelas 1. Oleh karena itu, jika kelas 1 adalah kelas yang penting, perlu dilakukan peningkatan lebih lanjut pada model atau penyesuaian strategi klasifikasinya.

# **Model5**
### Soal :
Jelaskan secara Singkat Model ke-5 yang digunakan!

In [ ]:
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

# Initialize the Support Vector Machine (SVM) model
model3 = SVC()

# Define the set of hyperparameters to be tested
params = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],  # Inverse regularization parameter
    'kernel': ['rbf', 'poly', 'sigmoid'],  # Kernel types
    'gamma': [0.001, 0.01, 0.1, 1, 'scale', 'auto'],  # Kernel parameter
}

# Perform GridSearchCV for hyperparameter tuning
grid = GridSearchCV(
             estimator=model3,  # Model to be tested
             param_grid=params,  # Set of hyperparameters to be tested
             scoring='accuracy',  # Evaluation metric used
             n_jobs=10,  # Number of CPU cores to use
             cv=3  # 3-fold cross-validation (meaning the model is iterated 3 times)
            )

# Train the model with the best combination of hyperparameters
grid.fit(X_train, y_train)

# Display the best hyperparameters found by GridSearchCV
best_params = grid.best_params_
best_params

{'C': 100, 'gamma': 'auto', 'kernel': 'rbf'}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.86      0.98      0.92      1983
           1       0.84      0.39      0.53       517

    accuracy                           0.86      2500
   macro avg       0.85      0.69      0.73      2500
weighted avg       0.86      0.86      0.84      2500


[[1944   39]
 [ 314  203]]

0.8588


## **Kesimpulan Model SVM**

Model Support Vector Machine (SVM) adalah salah satu model machine learning yang digunakan untuk masalah klasifikasi dan regresi. Model ini terutama dikenal dalam konteks klasifikasi. SVM bekerja dengan cara mencari hiperplane terbaik yang memisahkan dua kelas dalam ruang fitur. Keuntungan dari model SVM adalah kemampuannya untuk mengatasi masalah klasifikasi dalam ruang fitur yang kompleks dan mampu menangani data yang tidak linear. Kinerja model SVM dievaluasi dengan metrik-metrik seperti akurasi, precision, recall, dan F1-score.

1. Precision:

> - Precision adalah ukuran sejauh mana model memprediksi kelas positif dengan benar.
> - Untuk kelas 0 (label 0), precision adalah 0.86, yang berarti bahwa dari semua prediksi yang diklasifikasikan sebagai kelas 0, sekitar 86% benar-benar adalah kelas 0.
> - Untuk kelas 1 (label 1), precision adalah 0.84, yang berarti bahwa dari semua prediksi yang diklasifikasikan sebagai kelas 1, sekitar 84% benar-benar adalah kelas 1.

2. Recall (Sensitivity):

> - Recall adalah ukuran sejauh mana model berhasil mengidentifikasi semua kasus yang sebenarnya positif.
> - Untuk kelas 0 (label 0), recall adalah 0.98, yang mengindikasikan bahwa model sangat baik dalam mengidentifikasi data yang sebenarnya adalah kelas 0.
> - Untuk kelas 1 (label 1), recall adalah 0.39, yang menunjukkan bahwa model memiliki kesulitan dalam mengidentifikasi data yang sebenarnya adalah kelas 1.

3. F1-Score:

> - F1-score adalah rata-rata harmonik dari precision dan recall. Ini memberikan nilai tunggal yang mencakup baik precision maupun recall.
> - F1-score untuk kelas 0 adalah 0.92, dan untuk kelas 1 adalah 0.53.

4. Support:

> Support adalah jumlah data aktual yang termasuk dalam setiap kelas. Ada 1983 data dengan label 0 dan 517 data dengan label 1.

5. Accuracy:

> Akurasi model adalah 0.8588 atau sekitar 85.88%, yang berarti model memiliki tingkat keakuratan yang baik pada dataset ini.

6. Confusion Matrix:

> - Confusion matrix menunjukkan perbandingan antara prediksi model dengan kelas sebenarnya.
> - Diagonal utama dari matriks ini adalah true positive (1944) dan true negative (203), yang berarti model benar-benar memprediksi dengan benar kelas 0 sebanyak 1944 kali dan kelas 1 sebanyak 203 kali.
> - Diagonal yang tidak utama adalah false positive (39) dan false negative (314), yang mengindikasikan bahwa model melakukan beberapa kesalahan dalam mengklasifikasikan data kelas 1.

Kesimpulannya, model ini memiliki kinerja yang baik dalam mengidentifikasi data kelas 0 dengan tingkat precision dan recall yang tinggi. Namun, model memiliki kesulitan dalam mengidentifikasi data kelas 1, yang tercermin dalam recall yang lebih rendah. Akurasi model secara keseluruhan juga cukup baik.

# **Model6**
### Soal :
Jelaskan secara Singkat Model ke-6 yang digunakan!

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV

# Initialize the Decision Tree model
model6 = DecisionTreeClassifier()

# Define the set of hyperparameters to be tested
params = {
    'criterion': ['gini', 'entropy'],  # Splitting criterion
    'max_depth': [None, 10, 20, 30],  # Maximum depth of the tree
    'min_samples_split': [2, 5, 10],  # Minimum samples required to split a node
    'min_samples_leaf': [1, 2, 4],  # Minimum samples required at a leaf node
}

# Perform parameter tuning according to the required hyperparameters
grid = GridSearchCV(
             estimator=model6,
             param_grid=params,
             scoring='accuracy',
             n_jobs=10,  # Number of CPU cores used
             cv=10  # 10-fold cross-validation
            )

# Train the model with the best combination of hyperparameters
grid.fit(X_train, y_train)

# Display the best hyperparameters found by GridSearchCV
best_params = grid.best_params_
best_params

{'criterion': 'entropy',
 'max_depth': 10,
 'min_samples_leaf': 2,
 'min_samples_split': 10}

In [ ]:
#lakukan evaluasi
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

y_pred = grid.predict(X_test)

print(classification_report(y_test,y_pred))
print("")
print(confusion_matrix(y_test,y_pred))
print("")
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.87      0.95      0.91      1983
           1       0.70      0.48      0.57       517

    accuracy                           0.85      2500
   macro avg       0.79      0.71      0.74      2500
weighted avg       0.84      0.85      0.84      2500


[[1878  105]
 [ 271  246]]

0.8496


## **Kesimpulan Model Decision Tree**

Decision Tree adalah struktur pohon yang digunakan untuk mengambil keputusan dengan cara membagi dataset ke dalam subset yang lebih kecil berdasarkan fitur-fitur yang ada. Model Decision Tree bekerja dengan membagi dataset berulang-ulang berdasarkan fitur-fitur yang ada hingga mencapai simpul daun yang menghasilkan prediksi. Hasil dari evaluasi kinerja model Decision Tree ditampilkan dalam metrik seperti precision, recall, F1-score, dan akurasi.

1. Precision:

> - Precision adalah ukuran sejauh mana model memprediksi kelas positif dengan benar.
> - Untuk kelas 0 (label 0), precision adalah 0.87, yang berarti bahwa dari semua prediksi yang diklasifikasikan sebagai kelas 0, sekitar 87% benar-benar adalah kelas 0.
> - Untuk kelas 1 (label 1), precision adalah 0.70, yang berarti bahwa dari semua prediksi yang diklasifikasikan sebagai kelas 1, sekitar 70% benar-benar adalah kelas 1.

2. Recall (Sensitivity):

> - Recall adalah ukuran sejauh mana model berhasil mengidentifikasi semua kasus yang sebenarnya positif.
> - Untuk kelas 0 (label 0), recall adalah 0.95, yang mengindikasikan bahwa model sangat baik dalam mengidentifikasi data yang sebenarnya adalah kelas 0.
> - Untuk kelas 1 (label 1), recall adalah 0.48, yang menunjukkan bahwa model memiliki kesulitan dalam mengidentifikasi data yang sebenarnya adalah kelas 1.

3. F1-Score:

> - F1-score adalah rata-rata harmonik dari precision dan recall. Ini memberikan nilai tunggal yang mencakup baik precision maupun recall.
> - F1-score untuk kelas 0 adalah 0.91, dan untuk kelas 1 adalah 0.57.

4. Support:

> Support adalah jumlah data aktual yang termasuk dalam setiap kelas. Ada 1983 data dengan label 0 dan 517 data dengan label 1.

5. Accuracy:

> Akurasi model adalah 0.8496 atau sekitar 84.96%, yang berarti model memiliki tingkat keakuratan yang baik pada dataset ini.

6. Confusion Matrix:

> - Confusion matrix menunjukkan perbandingan antara prediksi model dengan kelas sebenarnya.
> - Diagonal utama dari matriks ini adalah true positive (1878) dan true negative (246), yang berarti model benar-benar memprediksi dengan benar kelas 0 sebanyak 1878 kali dan kelas 1 sebanyak 246 kali.
> - Diagonal yang tidak utama adalah false positive (105) dan false negative (271), yang mengindikasikan bahwa model melakukan beberapa kesalahan dalam mengklasifikasikan data kelas 1.
> - Kesimpulannya, model ini memiliki kinerja yang cukup baik dalam mengidentifikasi data kelas 0 dengan tingkat precision dan recall yang tinggi. Namun, model memiliki kesulitan dalam mengidentifikasi data kelas 1, yang tercermin dalam recall yang lebih rendah. Akurasi model secara keseluruhan juga cukup baik, tetapi perlu mempertimbangkan trade-off antara precision dan recall tergantung pada tujuan.

## Tarik Kesimpulan Model Mana yang terbaik beserta alasannya

In [ ]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Bagi data menjadi data pelatihan dan data pengujian
X_train,X_test,y_train,y_test = train_test_split(X_transform,y,test_size = 0.25,random_state = 123)

# Inisialisasi model-model yang akan diuji
models = [
    ("Random Forest", RandomForestClassifier()),
    ("Gradient Boosting", GradientBoostingClassifier()),
    ("Logistic Regression", LogisticRegression()),
    ("SVM", SVC()),
    ("K-Nearest Neighbors", KNeighborsClassifier()),
    ("Decision Tree", DecisionTreeClassifier())
]

# Evaluasi dan pilih model terbaik
best_model = None
best_accuracy = 0

for name, model in models:
    # Lakukan validasi silang dengan model
    scores = cross_val_score(model, X_train, y_train, cv=10, scoring='accuracy')
    avg_accuracy = scores.mean()

    # Cetak hasil validasi silang
    print(f"{name}: {avg_accuracy:.4f}")

    # Periksa apakah model ini lebih baik dari yang sebelumnya
    if avg_accuracy > best_accuracy:
        best_model = model
        best_accuracy = avg_accuracy

# Latih model terbaik dengan data pelatihan
best_model.fit(X_train, y_train)

# Lakukan prediksi dengan model terbaik pada data pengujian
y_pred = best_model.predict(X_test)

# Evaluasi model terbaik pada data pengujian
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy on test data: {accuracy:.4f}")

# Model terbaik adalah best_model
print("Best Model:", best_model)

Random Forest: 0.8611
Gradient Boosting: 0.8615
Logistic Regression: 0.8105
SVM: 0.8289
K-Nearest Neighbors: 0.8020
Decision Tree: 0.7915
Accuracy on test data: 0.8668
Best Model: GradientBoostingClassifier()


Jadi, berdasarkan hasil pengujian yang telah dilakukan, diperoleh bahwa nilai akurasi model yang paling tinggi adalah (1) Gradient Boosting dengan akurasi 86,4%, (2) Random Forest dengan akurasi 86,1%, dan (3) Support Vector Machine (SVM) dengan akurasi 85,8%.